### PR TITLE
Update http client config generation

### PR DESCRIPTION
**Pull request by Chuhaa**
## Purpose
Update http client config generation

## Goals
Fixes https://github.com/wso2-enterprise/choreo/issues/14786

## Documentation
Implemented design

- **Category1** (No Auth)
```ballerina
public type ConnectionConfig record {|
    # The HTTP version understood by the client
    http:HttpVersion httpVersion = http:HTTP_2_0;
    # Configurations related to HTTP/1.x protocol
    ClientHttp1Settings http1Settings?;
    # Configurations related to HTTP/2 protocol
    http:ClientHttp2Settings http2Settings?;
    # The maximum time to wait (in seconds) for a response before closing the connection
    decimal timeout = 60;
    # The choice of setting `forwarded`/`x-forwarded` header
    string forwarded = "disable";
    # Configurations associated with request pooling
    http:PoolConfiguration poolConfig?;
    # HTTP caching related configurations
    http:CacheConfig cache?;
    # Specifies the way of handling compression (`accept-encoding`) header
    http:Compression compression = http:COMPRESSION_AUTO;
    # Configurations associated with the behaviour of the Circuit Breaker
    http:CircuitBreakerConfig circuitBreaker?;
    # Configurations associated with retrying
    http:RetryConfig retryConfig?;
    # Configurations associated with inbound response size limits
    http:ResponseLimitConfigs responseLimits?;
    # SSL/TLS-related options
    http:ClientSecureSocket secureSocket?;
    # Proxy server related options
    http:ProxyConfig proxy?;
    # Enables the inbound payload validation functionality which provided by the constraint package. Enabled by default
    boolean validation = true;
|};

public isolated function init(string serviceUrl, ConnectionConfig config) returns error? {
   http:ClientConfiguration httpClientConfig = {httpVersion: config.httpVersion, timeout: config.timeout, forwarded: config.forwarded, poolConfig: config.poolConfig, compression: config.compression, circuitBreaker: config.circuitBreaker, retryConfig: config.retryConfig, validation:config.validation};
    do {
        if config.http1Settings is ClientHttp1Settings {
            ClientHttp1Settings settings = check config.http1Settings.ensureType(ClientHttp1Settings);
            httpClientConfig.http1Settings = {...settings};
        }
        if config.http2Settings is http:ClientHttp2Settings {
            httpClientConfig.http2Settings = check config.http2Settings.ensureType(http:ClientHttp2Settings);
        }
        if config.cache is http:CacheConfig {
            httpClientConfig.cache = check config.cache.ensureType(http:CacheConfig);
        }
        if config.responseLimits is http:ResponseLimitConfigs {
            httpClientConfig.responseLimits = check config.responseLimits.ensureType(http:ResponseLimitConfigs);
        }
        if config.secureSocket is http:ClientSecureSocket {
            httpClientConfig.secureSocket = check config.secureSocket.ensureType(http:ClientSecureSocket);
        }
        if config.proxy is http:ProxyConfig {
            httpClientConfig.proxy = check config.proxy.ensureType(http:ProxyConfig);
        }
    } on fail var e {
        return error graphql:RequestError("GraphQL Client Error", e);
    }
    graphql:Client clientEp = check new (serviceUrl, httpClientConfig);
    self.graphqlClient = clientEp;
}
```

- **Category2** (Only API Key Config)
```ballerina
public type ConnectionConfig record {|
    # The HTTP version understood by the client
    http:HttpVersion httpVersion = http:HTTP_2_0;
    # Configurations related to HTTP/1.x protocol
    ClientHttp1Settings http1Settings?;
    # Configurations related to HTTP/2 protocol
    http:ClientHttp2Settings http2Settings?;
    # The maximum time to wait (in seconds) for a response before closing the connection
    decimal timeout = 60;
    # The choice of setting `forwarded`/`x-forwarded` header
    string forwarded = "disable";
    # Configurations associated with request pooling
    http:PoolConfiguration poolConfig?;
    # HTTP caching related configurations
    http:CacheConfig cache?;
    # Specifies the way of handling compression (`accept-encoding`) header
    http:Compression compression = http:COMPRESSION_AUTO;
    # Configurations associated with the behaviour of the Circuit Breaker
    http:CircuitBreakerConfig circuitBreaker?;
    # Configurations associated with retrying
    http:RetryConfig retryConfig?;
    # Configurations associated with inbound response size limits
    http:ResponseLimitConfigs responseLimits?;
    # SSL/TLS-related options
    http:ClientSecureSocket secureSocket?;
    # Proxy server related options
    http:ProxyConfig proxy?;
    # Enables the inbound payload validation functionality which provided by the constraint package. Enabled by default
    boolean validation = true;
|};

public isolated function init(ApiKeysConfig apiKeysConfig, string serviceUrl, ConnectionConfig config) returns error? {
    http:ClientConfiguration httpClientConfig = {httpVersion: config.httpVersion, timeout: config.timeout, forwarded: config.forwarded, poolConfig: config.poolConfig, compression: config.compression, circuitBreaker: config.circuitBreaker, retryConfig: config.retryConfig, validation:config.validation};
    do {
        if config.http1Settings is ClientHttp1Settings {
            ClientHttp1Settings settings = check config.http1Settings.ensureType(ClientHttp1Settings);
            httpClientConfig.http1Settings = {...settings};
        }
        if config.http2Settings is http:ClientHttp2Settings {
            httpClientConfig.http2Settings = check config.http2Settings.ensureType(http:ClientHttp2Settings);
        }
        if config.cache is http:CacheConfig {
            httpClientConfig.cache = check config.cache.ensureType(http:CacheConfig);
        }
        if config.responseLimits is http:ResponseLimitConfigs {
            httpClientConfig.responseLimits = check config.responseLimits.ensureType(http:ResponseLimitConfigs);
        }
        if config.secureSocket is http:ClientSecureSocket {
            httpClientConfig.secureSocket = check config.secureSocket.ensureType(http:ClientSecureSocket);
        }
        if config.proxy is http:ProxyConfig {
            httpClientConfig.proxy = check config.proxy.ensureType(http:ProxyConfig);
        }
    } on fail var e {
        return error graphql:RequestError("GraphQL Client Error", e);
    }
    graphql:Client clientEp = check new (serviceUrl, httpClientConfig);
    self.graphqlClient = clientEp;
    self.apiKeyConfig = config.apiKeyConfig.cloneReadOnly();
}
```

**Category3** (Bearer token or Basic)
```ballerina
public type ConnectionConfig record {|
    # Configurations related to client authentication
    http:BearerTokenConfig auth;
    # The HTTP version understood by the client
    http:HttpVersion httpVersion = http:HTTP_2_0;
    # Configurations related to HTTP/1.x protocol
    ClientHttp1Settings http1Settings?;
    # Configurations related to HTTP/2 protocol
    http:ClientHttp2Settings http2Settings?;
    # The maximum time to wait (in seconds) for a response before closing the connection
    decimal timeout = 60;
    # The choice of setting `forwarded`/`x-forwarded` header
    string forwarded = "disable";
    # Configurations associated with request pooling
    http:PoolConfiguration poolConfig?;
    # HTTP caching related configurations
    http:CacheConfig cache?;
    # Specifies the way of handling compression (`accept-encoding`) header
    http:Compression compression = http:COMPRESSION_AUTO;
    # Configurations associated with the behaviour of the Circuit Breaker
    http:CircuitBreakerConfig circuitBreaker?;
    # Configurations associated with retrying
    http:RetryConfig retryConfig?;
    # Configurations associated with inbound response size limits
    http:ResponseLimitConfigs responseLimits?;
    # SSL/TLS-related options
    http:ClientSecureSocket secureSocket?;
    # Proxy server related options
    http:ProxyConfig proxy?;
    # Enables the inbound payload validation functionality which provided by the constraint package. Enabled by default
    boolean validation = true;
|};

public isolated function init(string serviceUrl, ConnectionConfig config) returns error? {
    http:ClientConfiguration httpClientConfig = {auth: config.auth, httpVersion: config.httpVersion, timeout: config.timeout, forwarded: config.forwarded, poolConfig: config.poolConfig, compression: config.compression, circuitBreaker: config.circuitBreaker, retryConfig: config.retryConfig, validation:config.validation};
    do {
        if config.http1Settings is ClientHttp1Settings {
            ClientHttp1Settings settings = check config.http1Settings.ensureType(ClientHttp1Settings);
            httpClientConfig.http1Settings = {...settings};
        }
        if config.http2Settings is http:ClientHttp2Settings {
            httpClientConfig.http2Settings = check config.http2Settings.ensureType(http:ClientHttp2Settings);
        }
        if config.cache is http:CacheConfig {
            httpClientConfig.cache = check config.cache.ensureType(http:CacheConfig);
        }
        if config.responseLimits is http:ResponseLimitConfigs {
            httpClientConfig.responseLimits = check config.responseLimits.ensureType(http:ResponseLimitConfigs);
        }
        if config.secureSocket is http:ClientSecureSocket {
            httpClientConfig.secureSocket = check config.secureSocket.ensureType(http:ClientSecureSocket);
        }
        if config.proxy is http:ProxyConfig {
            httpClientConfig.proxy = check config.proxy.ensureType(http:ProxyConfig);
        }
    } on fail var e {
        return error graphql:RequestError("GraphQL Client Error", e);
    }
    graphql:Client clientEp = check new (serviceUrl, httpClientConfig);
    self.graphqlClient = clientEp;
}
```

**Category4** (API Key Config + Other HTTP Authentication)
```ballerina
public type ConnectionConfig record {|
    # Configurations related to client authentication
    http:BearerTokenConfig auth;
    # The HTTP version understood by the client
    http:HttpVersion httpVersion = http:HTTP_2_0;
    # Configurations related to HTTP/1.x protocol
    ClientHttp1Settings http1Settings?;
    # Configurations related to HTTP/2 protocol
    http:ClientHttp2Settings http2Settings?;
    # The maximum time to wait (in seconds) for a response before closing the connection
    decimal timeout = 60;
    # The choice of setting `forwarded`/`x-forwarded` header
    string forwarded = "disable";
    # Configurations associated with request pooling
    http:PoolConfiguration poolConfig?;
    # HTTP caching related configurations
    http:CacheConfig cache?;
    # Specifies the way of handling compression (`accept-encoding`) header
    http:Compression compression = http:COMPRESSION_AUTO;
    # Configurations associated with the behaviour of the Circuit Breaker
    http:CircuitBreakerConfig circuitBreaker?;
    # Configurations associated with retrying
    http:RetryConfig retryConfig?;
    # Configurations associated with inbound response size limits
    http:ResponseLimitConfigs responseLimits?;
    # SSL/TLS-related options
    http:ClientSecureSocket secureSocket?;
    # Proxy server related options
    http:ProxyConfig proxy?;
    # Enables the inbound payload validation functionality which provided by the constraint package. Enabled by default
    boolean validation = true;
|};

public isolated function init(ConnectionConfig config, ApiKeysConfig apiKeysConfig, string serviceUrl) returns error? {
    http:ClientConfiguration httpClientConfig = {auth: config.auth, httpVersion: config.httpVersion, timeout: config.timeout, forwarded: config.forwarded, poolConfig: config.poolConfig, compression: config.compression, circuitBreaker: config.circuitBreaker, retryConfig: config.retryConfig, validation:config.validation};
    do {
        if config.http1Settings is ClientHttp1Settings {
            ClientHttp1Settings settings = check config.http1Settings.ensureType(ClientHttp1Settings);
            httpClientConfig.http1Settings = {...settings};
        }
        if config.http2Settings is http:ClientHttp2Settings {
            httpClientConfig.http2Settings = check config.http2Settings.ensureType(http:ClientHttp2Settings);
        }
        if config.cache is http:CacheConfig {
            httpClientConfig.cache = check config.cache.ensureType(http:CacheConfig);
        }
        if config.responseLimits is http:ResponseLimitConfigs {
            httpClientConfig.responseLimits = check config.responseLimits.ensureType(http:ResponseLimitConfigs);
        }
        if config.secureSocket is http:ClientSecureSocket {
            httpClientConfig.secureSocket = check config.secureSocket.ensureType(http:ClientSecureSocket);
        }
        if config.proxy is http:ProxyConfig {
            httpClientConfig.proxy = check config.proxy.ensureType(http:ProxyConfig);
        }
    } on fail var e {
        return error graphql:RequestError("GraphQL Client Error", e);
    }
    graphql:Client clientEp = check new (serviceUrl, httpClientConfig);
    self.graphqlClient = clientEp;
    self.apiKeysConfig = apiKeysConfig.cloneReadOnly();
}
```

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Related PRs
PRs for master branch
https://github.com/ballerina-platform/graphql-tools/pull/90
https://github.com/ballerina-platform/graphql-tools/pull/99

## Test environment
ballerina 2201.2.0

### CODE REVIEW

File: graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignatureWithApiKeysConfig.bal [hunk 1]
```diff
@@ -1 +1 @@
-(ApiKeysConfig apiKeysConfig, string serviceUrl, http:ClientConfiguration clientConfig = {}) returns graphql:ClientError?
+(ApiKeysConfig apiKeysConfig, string serviceUrl, ConnectionConfig config) returns graphql:ClientError?
```
Comment (RolandHewage) [inline graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignatureWithApiKeysConfig.bal:1]: Shouldn't we provide the default `{}` value for the `ConnectionConfig config`

File: graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignature.bal [hunk 2]
```diff
@@ -1 +1 @@
-(string serviceUrl, http:ClientConfiguration clientConfig = {}) returns graphql:ClientError?
+(string serviceUrl, ConnectionConfig config) returns graphql:ClientError?
```
Comment (RolandHewage) [inline graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignature.bal:1]: Shouldn't we provide the default `{}` value for the `ConnectionConfig config`

Comment (Chuhaa) [reply graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignature.bal]: Updated in https://github.com/ballerina-platform/graphql-tools/commit/1af52a4706e5087de8621c9e7ce382bbf445b16e

File: graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignatureWithApiKeysConfig.bal (hunk 1 above)
Comment (Chuhaa) [reply graphql-cli/src/test/resources/expectedGenCode/functionSignature/initFunctionSignatureWithApiKeysConfig.bal]: Updated in https://github.com/ballerina-platform/graphql-tools/commit/1af52a4706e5087de8621c9e7ce382bbf445b16e

**Review by RolandHewage — APPROVED** [review]

**Comment by RolandHewage** [comment]
LGTM. Code level changes are fine.

**Review by udda1996 — APPROVED** [review]

**Pull request merged**